src/manifest.c

```diff
@@ -171,11 +171,18 @@ static struct manifest *manifest_from_file(int version, char *component, bool he
 	struct manifest *manifest;
 	struct list *includes = NULL;
 	char *filename;
+	char *basedir;
+	uint64_t filecount = 0;
 	uint64_t contentsize = 0;
 	int manifest_hdr_version;
 	int manifest_enc_version;
 
-	string_or_die(&filename, "%s/%i/Manifest.%s", state_dir, version, component);
+	if (!statepath) {
+		basedir = state_dir;
+	} else {
+		basedir = statepath;
+	}
```

Comment:
Is your "statepath" always the same for a mix? If so, I would remove the statepath parameter and set basedir according to the value of is_mix (i.e. the global value, or a hardcoded value for the mix).